lib/AST/GenericSignature.cpp

```diff
@@ -814,16 +814,27 @@ ConformanceAccessPath GenericSignature::getConformanceAccessPath(
 
   // Local function to construct the conformance access path from the
   // requirement.
-  std::function<void(GenericSignature *, const RequirementSource *,
-                     ProtocolDecl *, Type)> buildPath;
-  buildPath = [&](GenericSignature *sig, const RequirementSource *source,
-                  ProtocolDecl *conformingProto, Type rootType) {
+  std::function<void(ArrayRef<Requirement>, const RequirementSource *,
+                     ProtocolDecl *, Type, ProtocolDecl *)> buildPath;
+  buildPath = [&](ArrayRef<Requirement> reqs, const RequirementSource *source,
+                  ProtocolDecl *conformingProto, Type rootType,
+                  ProtocolDecl *requirementSignatureProto) {
     // Each protocol requirement is a step along the path.
     if (source->isProtocolRequirement()) {
+      // If we're expanding for a protocol that has no requirement signature
+      // (yet) and have hit the penultimate step, this is the last step
+      // that would occur in the requirement signature.
+      if (!source->parent->parent && requirementSignatureProto) {
```

Comment:
Yeah, this is the fine distinction between the `usesRequirementSignature` bit in the requirement source and the protocol having a requirement signature. At this point in the code base, every protocol we encounter must have a requirement signature, because we're canonicalizing to that requirement signature.

However, the requirement sources we see might have been formed before that requirement signature was known, e.g., when there are recursive protocols. Thus, we have to canonicalize the path we create based on the now-known requirement signature. We pass `requirementSignatureProto` down when we're in that case.

The actual value of `requirementSignatureProto` isn't used; that's an artifact of my development process. I will replace it with a `bool` to indicate that we're canonicalizing a requirement source that didn't use a requirement signature to now use a requirement signature. Would it be clearer if we called this `bool` value `canonicalizingPath`?